flows/src/main/java/com/softwaremill/jox/flows/Flow.java

```diff
@@ -1296,139 +1204,220 @@ public Flow<T> alsoToTap(Sink<T> other) {
-        };
+    public Publisher<T> toPublisher(Scope scope) {
+        return new FromFlowPublisher<>(scope, last);
     }
 
+    // endregion
+
+    // region ByteFlow
+
+    public interface ByteChunkMapper<T> extends Function<T, ByteChunk> {}
+    public interface ByteArrayMapper<T> extends Function<T, byte[]> {}
+
     /**
-     * Writes content of this flow to an {@link java.io.OutputStream}.
+     * Converts a flow of `byte[]` or {@link ByteChunk} into a dedicated Flow type {@link ByteFlow}.
      *
-     * @param outputStream
-     *   Target `OutputStream` to write to. Will be closed after finishing the process or on error.
+     * @throws IllegalArgumentException if the flow does not contain `byte[]` or {@link ByteChunk} elements.
      */
-    public void runToOutputStream(OutputStream outputStream) throws Exception {
-        try {
-            last.run(t -> {
-                if (t instanceof byte[] chunk) {
-                    outputStream.write(chunk);
-                } else {
-                    throw new IllegalArgumentException("requirement failed: method can be called only on flow containing byte[]");
-                }
-            });
-            close(outputStream, null);
-        } catch (Exception e) {
-            close(outputStream, e);
-            throw e;
-        }
+    public final ByteFlow toByteFlow() {
+        return new ByteFlow(last);
     }
 
-    /** Writes content of this flow to a file.
-     *
-     * @param path
-     *   Path to the target file. If not exists, it will be created.s
+    /**
+     * Allows to convert given flow into a {@link ByteFlow} by providing a mapping function that converts elements of the flow into {@link ByteChunk}.
      */
-    public void runToFile(Path path) throws Exception {
-        if (Files.isDirectory(path)) {
-            throw new IOException("Path %s is a directory".formatted(path));
+    public ByteFlow toByteFlow(ByteChunkMapper<T> f) {
+        return new ByteFlow(map(f).last);
+    }
+
+    /**
+     * Allows to convert given flow into a {@link ByteFlow} by providing a mapping function that converts elements of the flow into `byte[]`.
+     */
+    public ByteFlow toByteFlow(ByteArrayMapper<T> f) {
+        return new ByteFlow(map(f).last);
+    }
+
+    /**
+     * Encodes a flow of `String` into a flow of bytes using UTF-8.
+     */
+    public ByteFlow encodeUtf8() {
+        Flow<ByteChunk> flow = map(s -> {
+            if (s instanceof String string) {
+                return ByteChunk.fromArray(string.getBytes(StandardCharsets.UTF_8));
+            }
+            throw new IllegalArgumentException("requirement failed: method can be called only on flow containing String");
+        });
+        return new ByteFlow(flow.last);
+    }
+
+    /**
+     * Subclass dedicated for flow containing ByteChunk
+     */
+    public static class ByteFlow extends Flow<ByteChunk> {
+
+        @SafeVarargs
+        private <T> ByteFlow(FlowStage<T> last, T... args) {
```

Comment:
ah I think I understand the trick now, the user is never supposed to provide the `T... args`, but the compiler inserts an empty array. But does it work, i.e., is that array not of type `Object[]`? especially that we call it from a generic function?

Reply:
Yes, that's the trick, you should get empty array of the type T. 
Looks like moving it to constructor doesn't work as expected. Indeed we get `Object[]` there, but it's working when it's used on `toByteFlow`, probably because `T` is a different generic type in `ByteFlow`. 
  
I'll move it as it was previously, passing class object explicitly.